executor/src/witgen/mod.rs

```diff
@@ -136,66 +127,134 @@ where
-        let (_, col) = values.iter_mut().find(|(n, _)| *n == name).unwrap();
-        *col = data;
+        let poly_id = *poly_ids
+            .iter()
+            .find(|&p| fixed.column_name(p) == name)
+            .unwrap();
+        columns[&poly_id] = data;
     }
-    values
+
+    // Map from column id to name
+    // We can't used `fixed` here, because the name would have the wrong lifetime.
+    let mut col_names = analyzed
+        .committed_polys_in_source_order()
+        .iter()
+        .map(|(p, _)| (p.id, p.absolute_name.as_str()))
+        .collect::<BTreeMap<_, _>>();
+
+    columns
+        .into_iter()
+        .map(|(id, v)| (col_names.remove(&id.id).unwrap(), v))
+        .collect()
+}
+
+fn zip_relevant<'a, T>(
+    row1: &'a ColumnMap<T>,
+    row2: &'a ColumnMap<T>,
+    relevant_mask: &'a [bool],
+) -> impl Iterator<Item = (&'a T, &'a T)> {
+    row1.values()
+        .zip(row2.values())
+        .zip(relevant_mask.iter())
+        .filter(|(_, &b)| b)
+        .map(|(v, _)| v)
 }
 
 /// Checks if the last rows are repeating and returns the period.
 /// Only checks for periods of 1, 2, 3 and 4.
-fn rows_are_repeating<T: PartialEq>(values: &[&(&str, Vec<T>)]) -> Option<usize> {
-    if values.is_empty() {
+fn rows_are_repeating<T: PartialEq>(
+    rows: &[ColumnMap<T>],
+    relevant_mask: &[bool],
+) -> Option<usize> {
+    if rows.is_empty() {
         return Some(1);
-    } else if values[0].1.len() < 4 {
+    } else if rows.len() < 4 {
         return None;
     }
+
+    let len = rows.len();
     (1..=3).find(|&period| {
-        values.iter().all(|(_name, value)| {
-            let len = value.len();
-            (1..=period).all(|i| value[len - i - period] == value[len - i])
+        (1..=period).all(|i| {
+            zip_relevant(&rows[len - i - period], &rows[len - i], relevant_mask)
+                .all(|(a, b)| a == b)
         })
     })
 }
 
 /// Data that is fixed for witness generation.
 pub struct FixedData<'a, T> {
     degree: DegreeType,
-    fixed_col_values: Vec<&'a Vec<T>>,
-    fixed_cols: Vec<&'a PolynomialReference>,
-    witness_cols: &'a Vec<WitnessColumn<'a, T>>,
+    fixed_cols: ColumnMap<FixedColumn<'a, T>>,
+    witness_cols: ColumnMap<WitnessColumn<'a, T>>,
 }
 
 impl<'a, T> FixedData<'a, T> {
     pub fn new(
         degree: DegreeType,
-        fixed_col_values: Vec<&'a Vec<T>>,
-        fixed_cols: Vec<&'a PolynomialReference>,
-        witness_cols: &'a Vec<WitnessColumn<'a, T>>,
+        fixed_cols: ColumnMap<FixedColumn<'a, T>>,
+        witness_cols: ColumnMap<WitnessColumn<'a, T>>,
     ) -> Self {
         FixedData {
             degree,
-            fixed_col_values,
             fixed_cols,
             witness_cols,
         }
     }
 
-    fn witness_cols(&self) -> impl Iterator<Item = &WitnessColumn<T>> {
-        self.witness_cols.iter()
+    fn witness_map<V: Clone>(&self, initial_value: V) -> ColumnMap<V> {
```

Comment:
Do we ever use this function with anything else than `None`? If so, I'd rename to `witness_map_with`, if I'd remove the argument.

Reply:
It's used once with `vec![]`, where the map actually maps to entire columns. In #412, I have the concept of a `Row` which I guess should use a column map internally and will replace the `ColumnMap<Option<T>>` and `ColumnMap<Option<RangeConstraint>>`. That will make the distinction clearer.

Reply:
For now I renamed it!